### PR TITLE
Add stock location through ps_stock_available table

### LINKED ISSUE #10113: [BOOM-3885] Stock Location field missing
**NicholasIT**
```
Hi,
i've read about other old stock location requests and you're still planning to develop advanced stock management but i ask it again becouse we [i guess not just us but other merchants with big inventory] just need
a temporary field, anywhere in product page, where we can write the position of the product on the shelfs of our warehouse [for example "A-10"]
I guess this temporary field will help so much even advanced stock management is not complete at the moment.
Thanks for your attention and patience,
```

**PrestaShop team :**

In order to put stock location field back we should put it on several places and change some text (the field will be optional like in 1.6) :  

1) Where do we put the editable field ? :  

if the product has no combination
*Product Page > Quantity Tab*

Rename "Stock Alerts" to "Stock" and add a "Stock location" field

nocombinaison-location

if the product has combination(s) 
*Combination tab> Edit combination (pencil)*

Rename "Stock Alerts" to "Stock" and add a "Stock location" field 

combinaison-location

2) Where do we put the information ? 

*Order page> Order tab > Order selection*

At the bottom of the page, on the product, the information of stock location will show up.

If we click on edit (pencil), the stock location will be uneditable

order-stock-location

### PR DESCRIPTION
**Pull request by matks**
⚠️ PR mainly approved, only last commit https://github.com/PrestaShop/PrestaShop/pull/10422/commits/1abf49c9400dff20dac1a76d8ea410084f98f27b needs review

| Questions     | Answers
| ------------- | -------------------------------------------------------
| Branch?       | develop
| Description?  | Restore the use of the 'stock location' product field
| Type?         | new feature
| Category?     | BO
| BC breaks?    | no
| Deprecations? | no
| Fixed ticket? | Fixes #10113
| How to test?  |The expected behavior is described in the issue

---
This change is [Reviewable](https://reviewable.io/reviews/prestashop/prestashop/10422)

### CODE REVIEW

File: classes/order/Order.php
Comment (mickaelandrieu) [inline classes/order/Order.php]: this will break the compatibility: you can create a new protected method and deprecate this one if you think the naming of the former one is bad.

File: classes/stock/StockAvailable.php
Comment (mickaelandrieu) [inline classes/stock/StockAvailable.php:58]: :+1:

File: classes/stock/StockAvailable.php
Comment (mickaelandrieu) [inline classes/stock/StockAvailable.php]: We can do better here:

```php
if (
    false === Validate::isUnsignedId($id_product)
    || false === Validate::isUnsignedId($id_shop)
   || false === Validate::isUnsignedId($id_product_attribute)
   || false === Validate::isString($location)
) {
    throw new SomeSpecializedException('Some meaningful message with serialized data passed');
}
```

And you can remove almost all cast operations you're doing, also this will remove unexpected weird behaviors that can come from wrong data passed to this function.

File: classes/stock/StockAvailable.php
Comment (mickaelandrieu) [inline classes/stock/StockAvailable.php]: Same here, I think we should throw an exception instead :) it can be a simple InvalidArgumentException but avoid missing multiple return types to manage wrong inputs. Also, you can validate other fields.

File: classes/stock/StockAvailable.php
Comment (mickaelandrieu) [inline classes/stock/StockAvailable.php:764]: $id_shop should be default evaluated to 0, same for $id_product_attribute.

File: classes/stock/StockAvailable.php
Comment (mickaelandrieu) [inline classes/stock/StockAvailable.php]: here, you can do the trick to stick with the old API:

```php
// StockAvailable::addSqlShopRestriction expects Shop id to be null if not valid.
if ($id_shop !== 0) {
    $id_shop = null;
}
```

File: src/Adapter/Product/AdminProductWrapper.php
Comment (mickaelandrieu) [inline src/Adapter/Product/AdminProductWrapper.php]: you don't need to cast to integer if it comes from Product.

File: src/Adapter/Product/ProductDataProvider.php
Comment (mickaelandrieu) [inline src/Adapter/Product/ProductDataProvider.php]: I think you should use a default int value instead, as we already know 0 is the default value :+1:

File: src/PrestaShopBundle/Controller/Admin/ProductController.php
Comment (mickaelandrieu) [inline src/PrestaShopBundle/Controller/Admin/ProductController.php]: how about moving it into the parent class?

**Review by mickaelandrieu — DISMISSED** [review]

File: classes/order/Order.php
Comment (matks) [reply classes/order/Order.php]: First I considered adding a new protected method and keeping this one but the behavior change is too important, I think it is better to embark it inside the existing function so that developers using it do not have "lose" the feature :) 

Incoming in next commit

File: classes/stock/StockAvailable.php
Comment (matks) [reply classes/stock/StockAvailable.php:764]: Would it introduce some magic numbers inside this function ? I know in PrestaShop ID 0 is the default one. However adding it as default value would hard-code this rule into the code. What if one day we want to move to uuid SQL IDs ? Even if we extend StockAvailable, the value `0` will be hardcoded into the function signature.

So ... what about `null` as default value always :) ?

Comment (mickaelandrieu) [reply classes/stock/StockAvailable.php:764]: if we move from ids to uuid we will have bigger problems as we expected integer or null here and we will change the functions to use some Uuid objects, isn't it?

It's only 1 implementation, I don't think we will break it in 1.7.x release.

Anytime, when we can do it I want only 1 possible type of input (if we were able to use PHP 7.1 already, I could be fine with ?int but ... :'( ).

Other option: check if $id_product_attribute is a int if it's not a null value, wdyt?

Comment (PierreRambaud) [reply classes/stock/StockAvailable.php:764]: Even if we use only one type, we are able to set a value to `null`. Which is logic to have when there is no id attached to the object, because 0 is an id :)
If we move to uuid or anything else, it's easier to test if value is null instead of checking if it's numeric and not 0 (and it costs more resources).

```php
<?php

class A
{
    public function __construct(int $a)
    {
        var_dump($a);
    }
}

new A();
new A(22);
new A('azerty');
```

Will produce:
```bash
$ php7.2 test.php 
/home/prambaud/test.php:7:
NULL
/home/prambaud/test.php:7:
int(22)
PHP Fatal error:  Uncaught TypeError: Argument 1 passed to A::__construct() must be of the type integer or null, string given, called in /home/prambaud/test.php on line 13 and defined in /home/prambaud/test.php:5
Stack trace:
#0 /home/prambaud/test.php(13): A->__construct('azerty')
#1 {main}
  thrown in /home/prambaud/test.php on line 5
```

**Comment by matks** [comment]
@mickaelandrieu PR feedbacks have been implemented in https://github.com/PrestaShop/PrestaShop/pull/10422/commits/5b87438a839613641fb63f48f6a17fabd791fffa ;)

**Comment by eternoendless** [comment]
Code looks good, but tests are failing

**Comment by matks** [comment]
@eternoendless I saw 😢 I fixed it by removing some validations in `StockAvailable`

According to [AdminModelAdapterTest](https://github.com/PrestaShop/PrestaShop/blob/develop/tests/PrestaShopBundle/Model/Product/AdminModelAdapterTest.php#L136) we must accept `string` values for `$id_product_attribute` and `$id_product` in [getLocation()](https://github.com/PrestaShop/PrestaShop/pull/10422/files#diff-6a810cff1b3b082a1b2a3f1dbca30eeeR764)

Fixed by https://github.com/PrestaShop/PrestaShop/pull/10422/commits/9eae1b1267fb1093778bdbb02e2c9a3ba67c930d, should fix Travis build

**Review by PierreRambaud — DISMISSED** [review]

File: classes/stock/StockAvailable.php
Comment (mickaelandrieu) [inline classes/stock/StockAvailable.php:384]: isn't weird to throw an exception if the data is actually valid?

**Comment by mbadrani** [comment]
when I fill stock location like this

image

... And I access to order page, it display the number of quantity instead of stock location value

image

**Comment by matks** [comment]
@mbadrani Can you clear your cache on this very page (in chrome, open "developer tools", hold click on the "refresh page button" to make a menu appear and then select "clear cache and perform a forced actualisation of the page" ?

On your screenshot I see the row elements are dashed so I think is a an outdated CSS issue

**Comment by mbadrani** [comment]
After force refreshing and clear the cache I can reproduce this weired behaviour

**Comment by matks** [comment]
@mbadrani I found an issue with customized data indeed 😉 nice catch !

⚠️ Last commit https://github.com/PrestaShop/PrestaShop/pull/10422/commits/1abf49c9400dff20dac1a76d8ea410084f98f27b should fix it, waiting for review now

**Review by Quetzacoalt91 — DISMISSED** [review]

**Review by jolelievre — DISMISSED** [review]
good job!

**Comment by mbadrani** [comment]
@matks After this fix I have another bug below
https://drive.google.com/file/d/1eB-OvunAIy-46uiOKDtODnrJg8FBUyZ3/view?usp=sharing

**Comment by PierreRambaud** [comment]
I think it's not a bug due to his modification because I sometimes have the same bug on develop.

**Comment by Quetzacoalt91** [comment]
We never know, if you look at the symfony debug bar on the video, it seems an error 500 occurred (the icon becomes red).

@mbadrani can you please hover this icon and click on the sha1 related to the call in error? You will get the details.

**Comment by matks** [comment]
Again nice catch @mbadrani : the bug happens if you empty the "stock location" field. This was considered a not valid dataset by the code. I fixed:
- the usecase
- now even if the dataset is not valid, the BO does not freeze the way it does in your video

**Technical insights:**
Because of https://github.com/PrestaShop/PrestaShop/pull/10422/commits/5b87438a839613641fb63f48f6a17fabd791fffa#diff-6a810cff1b3b082a1b2a3f1dbca30eeeR384 `null` data was not accepted anymore as a valid input and was consequently rejected using a `\InvalidArgumentException`. However since that was an ajax call, JSON was expected although it was the HTML symfony error page that was returned, causing JS to crash.

**Comment by mbadrani** [comment]
@PierreRambaud it's not a random bug it has been easily reproduced by @matks

**Review by jolelievre — APPROVED** [review]
nice fix!

**Comment by mbadrani** [comment]
Well done guys!

**Comment by eternoendless** [comment]
Woot! 💪 
Let's merge this once travis is green

**Comment by Quetzacoalt91** [comment]
Thank you @matks

**Comment by 123monsite-regis** [comment]
hi,

about this, can we undestand that "public $location;" inside Combination Object definition is now obsolete and should be migrated when updating shop ?
i talk about : https://github.com/PrestaShop/PrestaShop/blob/develop/classes/Combination.php#L40
`    public $location;`


in fact some paid modules still use this field whith the same label, it's quite confusing.


regards

**Pull request merged**
